### PR TITLE
Add V0 tensor layout generation

### LINKED ISSUE #572: Implement legal Op configuration generation
Implement legal Op configuration generation per [Optimizer design spec](https://tenstorrent.sharepoint.com/:w:/s/Software/EXMXTtdy5XhFmIAn4sJP0MIB58qBMsr9W1fsLIOeCKiYvA?e=gVSc6Z). This includes generation of valid shard specification for supported ops.

### PR DESCRIPTION
**Pull request by odjuricicTT**
A first implementation of generating possible op configurations. Starting with generating a few sharded tensor layouts for now. Much of the details here are incomplete and this is meant to unblock further optimizer and runtime development.

Waiting on #541 in order to add appropriate TensorMemoryLayout before merge. 

Closes #572

### CODE REVIEW

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 1]
```diff
@@ -36,17 +41,71 @@ bool LegalGridAnalysis::applyOverrides() {
...
+  LayoutAttr l1Interleaved =
+      layout.withMemorySpace(op->getContext(), MemorySpace::DeviceL1);
+  analysisResult.push_back(l1Interleaved);
+
+  // L1 Sharded
+  LayoutAttr shardedBase =
+      layout.withMemorySpace(op->getContext(), MemorySpace::DeviceL1);
+
+  // Block Sharded
+  for (auto width = 2; width <= analysisInput.maxGrid.getShape()[0]; ++width) {
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:71]: Maybe start with max_width/2. Same for height. Are you sure all these have chance to be valid, to you need to check/factor tensor shape first?

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 2]
```diff
@@ -36,17 +41,71 @@ bool LegalGridAnalysis::applyOverrides() {
...
+  // No grid is set since the tensor is not sharded.
+  // TODO: Is this a viable solution or should we have a grid?
+  LayoutAttr dram =
+      layout.withMemorySpace(op->getContext(), MemorySpace::DeviceDRAM);
+  analysisResult.push_back(dram);
+
+  // L1 Interleaved (same as above)
+  LayoutAttr l1Interleaved =
+      layout.withMemorySpace(op->getContext(), MemorySpace::DeviceL1);
+  analysisResult.push_back(l1Interleaved);
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:64]: Is there any attribute that makes this one interleaved compared to sharded L1s below?

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 3]
```diff
@@ -36,17 +41,71 @@ bool LegalGridAnalysis::applyOverrides() {
...
+
+  // L1 Sharded
+  LayoutAttr shardedBase =
+      layout.withMemorySpace(op->getContext(), MemorySpace::DeviceL1);
+
+  // Block Sharded
+  for (auto width = 2; width <= analysisInput.maxGrid.getShape()[0]; ++width) {
+    for (auto height = 2; height <= analysisInput.maxGrid.getShape()[1];
+         ++height) {
+      analysisResult.push_back(shardedBase.withGrid(
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:74]: There is a shardSpec inside LayoutAttr is it also automatically updated?

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 4]
```diff
@@ -36,17 +41,71 @@ bool LegalGridAnalysis::applyOverrides() {
...
+
+  // Filter layouts based on output tensor legality for current op.
+  analysisResult.erase(
+      std::remove_if(analysisResult.begin(), analysisResult.end(),
+                     [this](LayoutAttr layout) {
+                       return !mock_is_output_tensor_legal_for_op(op, layout);
+                     }),
+      analysisResult.end());
+
+  // TODO: Potetialy filter out tensors that dont fit into L1 at all.
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:109]: It would be nice to have this for sharded ones.

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 5]
```diff
@@ -36,17 +41,71 @@ bool LegalGridAnalysis::applyOverrides() {
...
+          GridAttr::get(op->getContext(), {width, height})));
+    }
+  }
+
+  auto numCores =
+      analysisInput.maxGrid.getShape()[0] * analysisInput.maxGrid.getShape()[1];
+  // Height Sharded
+  // TODO: Missing affine mapping to actual grid.
+  // TODO: Can we have every shape of 1d grid? Probably not, need to check what
+  // is divisible by grid sides.
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:85]: I recommended above going down to 50% of grid usage.

**Review by nobradovictt — CHANGES_REQUESTED** [review]

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 2 above)
Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:64]: No, the only difference is that interleaved does not have a gird set. Should we make this explicit in LayoutAttr?

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 3 above)
Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:74]: I cannot seem to find it. Can you point me to what you are referring to?

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 4 above)
Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:109]: Ok, will treat this as a nice to have.

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 3 above)
Comment (nobradovictt) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:74]: Take a look at llvm::SmallVector<int64_t> LayoutAttr::getShardShape()

Comment (nobradovictt) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:74]: `                        AttrParameter<"MemRefType", "A memref that describes the physical footprint allocation of the shard. It must also have a shape with rank equal to grid.">:$memref);`

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 2 above)
Comment (nsmithtt) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:93]: @jnie-TT is adding flatbuffer support.  Once that's in I think we can just add a new layout attribute to capture this.  Or maybe we should specialize a TTNN layout class?  So far it's just 1 enum, so specializing might be overkill, metal backend can just assert that this enum is always set to `None` or something.

Comment (nobradovictt) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:93]: Great, thanks! Yeah lets go with new layout attribute.

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 4 above)
Comment (nobradovictt) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:138]: Actually, don't do this. :) We can still decide later to split the tensor for pipeline op, so this will be up to sharding policy to decide.

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 3 above)
Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:103]: Yes, it's automatically updated.

**Comment by odjuricicTT** [comment]
Waiting on #541 to land in order to set appropriate TensorMemoryLayout.

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 6]
```diff
@@ -6,11 +6,44 @@
...
+
+  return (MTiles % gridR == 0) && (KTIles % gridC == 0);
+}
+
 bool LegalGridAnalysis::applyOverrides() {
   // Lookup grid size overrides based on location information for current
   // operation.
   //
 
+  // TODO(odjuricic): We may need to infer shard type.
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:46]: Yeah override needs to be extended.

Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:46]: Do we want all LayoutAttr params to be overridable?

Comment (nobradovictt) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:46]: Because of linking with TT-explorer most likely yes, but let's find some good enough starting point.

Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:46]: Ok, will add this as separate issue and tackle after this.

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 7]
```diff
@@ -36,17 +70,95 @@ bool LegalGridAnalysis::applyOverrides() {
...
-  //
+  // A first incomplete implementation of the LegalGridAnalysis.
+  // This implementation is a placeholder and is meant to just enable testing of
+  // other components.
+
+  // Skip mlir ops.
+  if (not llvm::isa<TTIROp>(op)) {
+    return;
+  }
+  // Skip operations that don't have output tensors.
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:81]: Can this be used instead:
` if (op->getNumResults() == 0)`

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 8]
```diff
@@ -36,17 +70,95 @@ bool LegalGridAnalysis::applyOverrides() {
...
 void LegalGridAnalysis::analysisImplementation() {
-  // Placeholder, needs to be implemented. Go through all the grid sizes and
-  // check if they are legal based on tensor type and device/chip attributes.
-  // For now result of analysis is maximum supported grid size.
-  //
+  // A first incomplete implementation of the LegalGridAnalysis.
+  // This implementation is a placeholder and is meant to just enable testing of
+  // other components.
+
+  // Skip mlir ops.
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:77]: Nit: Update comment to: Process only TTIR ops.

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 7 above)
Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:81]: Yes, except for ToLayoutOp which has an output.

File: test/ttmlir/Dialect/TTIR/test_grid_set.mlir [hunk 9]
```diff
@@ -3,7 +3,7 @@
 module attributes {} {
   func.func @forward(%arg0: tensor<64x128xf32>, %arg1: tensor<64x128xf32>) -> tensor<64x128xf32> {
     %0 = tensor.empty() : tensor<64x128xf32>
-    // CHECK: #[[LAYOUT_1:.*]] = #tt.layout<(d0, d1) -> (d0, d1), undef, <8x8>, memref<8x16xf32, #dram>, interleaved>
+    // CHECK: #[[LAYOUT_1:.*]] = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<64x128xf32, #dram>, interleaved>
```
Comment (nobradovictt) [inline test/ttmlir/Dialect/TTIR/test_grid_set.mlir:6]: Why is this changed to 1x1?

File: include/ttmlir/Dialect/TTIR/Analysis/LegalGridAnalysis.h [hunk 10]
```diff
@@ -15,6 +15,7 @@ struct LegalGridAnalysisInput {
   ChipDescAttr chipDesc;
   GridAttr maxGrid;
   RankedTensorType tensorType;
+  int64_t maxShardedGrids = 64;
```
Comment (nobradovictt) [inline include/ttmlir/Dialect/TTIR/Analysis/LegalGridAnalysis.h:18]: constexpr

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 11]
```diff
@@ -3,14 +3,48 @@
...
+
+  return (MTiles % gridR == 0) && (KTIles % gridC == 0);
+}
+
 bool LegalGridAnalysis::applyOverrides() {
   // Lookup grid size overrides based on location information for current
   // operation.
   //
 
+  // TODO(odjuricic): Need to override all params, not just grid size.
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:47]: Is this for followup change? File an issue under Optmizier if so.

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp [hunk 12]
```diff
@@ -36,17 +70,108 @@ bool LegalGridAnalysis::applyOverrides() {
...
       mlir::cast<RankedTensorType>(op->getResult(0).getType());
   LayoutAttr layout = mlir::cast<LayoutAttr>(tensorType.getEncoding());
-  llvm::ArrayRef<int64_t> tensorShape = tensorType.getShape();
 
-  analysisResult.push_back(layout.withGrid(
-      op->getContext(), tensorShape,
-      GridAttr::get(op->getContext(), analysisInput.maxGrid.getShape())));
+  // DRAM
+  // No grid is set since the tensor is not sharded.
+  // TODO(odjuricic): We need to set grid here since it will be used as the
```
Comment (nobradovictt) [inline lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:96]: File an issue, mention special case where grid matches number of DRAM banks.

**Review by nobradovictt — APPROVED** [review]

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 11 above)
Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:47]: Yes, there is an issue filed.

File: include/ttmlir/Dialect/TTIR/Analysis/LegalGridAnalysis.h (hunk 10 above)
Comment (odjuricicTT) [reply include/ttmlir/Dialect/TTIR/Analysis/LegalGridAnalysis.h:18]: I was intending for this to be a passable param. Will set to constexpr for now and change when needed.

Comment (nobradovictt) [reply include/ttmlir/Dialect/TTIR/Analysis/LegalGridAnalysis.h:18]: Indeed it should be passed in as param, make it as part of override issue in followup change.

File: lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp (hunk 12 above)
Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:96]: Related to the 1x1 comment above, this needs to be updated once we know how runtime and TTNN treat this param.

Comment (odjuricicTT) [reply lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp:96]: Filed an issue.

File: test/ttmlir/Dialect/TTIR/test_grid_set.mlir (hunk 9 above)
Comment (nobradovictt) [reply test/ttmlir/Dialect/TTIR/test_grid_set.mlir:6]: @nsmithtt @jnie-TT Do we have any plans on how to infer computation grid for OPs which operate with DRAM data?

Comment (odjuricicTT) [reply test/ttmlir/Dialect/TTIR/test_grid_set.mlir:6]: This seems to be different on an op per op basis and is discussed in #450, #559 and is being addressed in #605 .
